go/mysql/binlog_event_compression.go

```diff
@@ -192,78 +230,126 @@ func (tp *TransactionPayload) read(data []byte) error {
-		if pos+eventLen > decompressedPayloadLen {
-			return vterrors.New(vtrpcpb.Code_INTERNAL,
-				fmt.Sprintf("[BUG] event length of %d at pos %d in decompressed transaction payload is beyond the expected payload length of %d",
-					eventLen, pos, decompressedPayloadLen))
+		if bytesRead != headerLen {
+			return nil, vterrors.Errorf(vtrpcpb.Code_INTERNAL, "[BUG] expected header length of %d but only read %d bytes",
+				headerLen, bytesRead)
 		}
-		eventData := decompressedPayload[pos : pos+eventLen]
-		ble := NewMysql56BinlogEvent(eventData)
-		tp.Events = append(tp.Events, ble)
-
-		pos += eventLen
+		eventLen := int64(binary.LittleEndian.Uint32(header[BinlogEventLenOffset:headerLen]))
+		eventData := make([]byte, eventLen)
+		copy(eventData, header) // The event includes the header
+		bytesRead, err = io.ReadFull(tp.reader, eventData[headerLen:])
+		if err != nil && err != io.EOF {
+			return nil, vterrors.Wrap(err, "error reading binlog event data from uncompressed transaction payload")
+		}
+		if int64(bytesRead+headerLen) != eventLen {
+			return nil, vterrors.Errorf(vtrpcpb.Code_INTERNAL, "[BUG] expected binlog event length of %d but only read %d bytes",
+				eventLen, bytesRead)
+		}
+		return NewMysql56BinlogEvent(eventData), nil
 	}
-
 	return nil
 }
 
-// Decompress the payload.
-func (tp *TransactionPayload) decompress() ([]byte, error) {
-	if len(tp.Payload) == 0 {
-		return []byte{}, vterrors.New(vtrpcpb.Code_INVALID_ARGUMENT, "cannot decompress empty payload")
+// decompress decompresses the payload. If the payload is larger than
+// zstdInMemoryDecompressorMaxSize then we stream the decompression via
+// the package's pool of zstd.Decoders, otherwise we use in-memory
+// buffers with the package's concurrent statelessDecoder.
+// In either case, we setup the reader that can be used within the
+// iterator to read the events one at a time from the decompressed
+// payload in GetNextEvent().
+func (tp *TransactionPayload) decompress() error {
+	if len(tp.payload) == 0 {
+		return vterrors.New(vtrpcpb.Code_INVALID_ARGUMENT, "cannot decompress empty compressed transaction payload")
 	}
-	var (
-		decompressedBytes []byte
-		err               error
-	)
-
-	// Switch to slower but less memory intensive stream mode for larger payloads.
-	if tp.UncompressedSize > zstdInMemoryDecompressorMaxSize {
-		in := bytes.NewReader(tp.Payload)
-		streamDecoder, err := zstd.NewReader(in)
-		if err != nil {
-			return nil, err
-		}
-		defer streamDecoder.Close()
-		out := io.Writer(&bytes.Buffer{})
-		_, err = io.Copy(out, streamDecoder)
-		if err != nil {
-			return nil, err
+
+	// Switch to slower but less memory intensive stream mode for
+	// larger payloads.
+	if tp.uncompressedSize > zstdInMemoryDecompressorMaxSize {
+		in := bytes.NewReader(tp.payload)
+		streamDecoder, ok := statefulDecoderPool.Get().(*zstd.Decoder)
+		if streamDecoder == nil || !ok {
```

Comment:
`ok` will always be `true` when `streamDecoder != nil`. You have determined the type in your implementation of `New()`, and so there can't be any other type. Doesn't hurt to check here as you already need to check for `nil` value, so this is just a nit comment.

Reply:
Yeah, it's not needed (just habit, I suppose). I removed it here: https://github.com/vitessio/vitess/pull/16328/commits/d5cbf05d3b0269a36a225b8b13b1952038c7f689